auth_core/pkce.py

```diff
@@ -0,0 +1,50 @@
+import base64
+import hashlib
+import httpx
+import logging
+import secrets
+
+from urllib.parse import urljoin
+
+from .token_data import TokenData
+
+
+logger = logging.getLogger("gel_auth_core")
+
+
+class PKCE:
+    def __init__(self, verifier: str, *, base_url: str):
+        self.base_url = base_url
+        self.verifier = verifier
+        self.challenge = (
```

Comment:
```suggestion
        self._verifier = verifier
        self._challenge = (
```

Let's make them semi-private, and expose read-only properties, because `challenge` is derived from `verifier`.